text/0021-versioning-releases.md

```diff
@@ -0,0 +1,26 @@
+# Releasing and Versioning of Openmrs SingleSpa Apps
+- Start Date: 2020/01/21
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/1
+
+## Decision
+Openmrs spa applications should have releases that are versioned so that each implementor can have control on when to upgrade to the latest changes. 
+
+The releases should have a maintenance branch for bug fixes and patches so that implementors running a particular version of an app can benefit from recent bug fixes. 
```

Comment:
At which level should these maintenance branches be maintained? Should a SPA maintain a branch for every minor version or only for major versions?

Reply:
Should this RFC make a decision on how bugfixes are addressed vis-a-vis more recent versions? I.e., should we favour bugfixes against 2.x which are backported to 1.x (this is something like the current OpenMRS model) or bugfixes against 1.x which are "forward-ported" to 2.x (this involves more work for the maintainers, but makes it easier for implementors to provide bugfixes).

Reply:
If we are going to do release branching, it should only be for major versions.

Reply:
I think we should have release branches for both, as the whole point is to allow implementors to decide when to upgrade to see new features, as well as allow them to benefit from bug-fixes.

Reply:
The alternative will be to do feature flags, which I think will get messy for an open-source project.

Reply:
The kinds of features that people would generally want to disable should be toggled in the config.

For everything else, organizations should ask themselves -- "are we really that special?" And if so, they should introduce a new config parameter.

Backporting bug fixes to every minor version:
- is not justifiable based on current adoption patterns
- will be a huge pain for developers
- will discourage us from releasing, knowing that every release is a new maintenance nightmare we will have to live with until it is EOL'd
- will reduce forward velocity on MFE before it has even had the opportunity to prove itself, at a time when organizations still consider dedicating devs to MFE a risky proposition

For major versions, there is much more of an argument to be made for release branching, since the definition of a major version is one that introduces breaking changes -- i.e., the upgrade requires some effort, and organizations often take a while to make that happen.

Reply:
If it was library that we are building, then what you are saying makes perfect sense. These front-end features are client-facing, users interact with them, there's training that goes with it, and decisions about workflows that are made by various stakeholders before a feature is introduced. We need to allow an organisation to benefit from bug-fixes and patches, before they are ready to let their users use a new feature or a modification of an existing feature. Effort to back-port bugs to certain release-branches could be left to such an organisation. Configs to support 'timing of release of a certain feature to a certain group of users' will require coordination each release cycle, and effort to maintain anyway, and can get ugly for an open-source project. 
All this said, I think it would be safer to allow creation of these minor release branches on demand, and leaving the responsibility of maintaining it to an organisation that needs it badly. Restricting it will only promote organisations to create forks when they get cornered with a bug that affects them.

Reply:
We still need the configs that are there by design, to allow an application to be configured to an organisation's  needs, but requiring that every feature to have these configs for the purposes of release adds complexities to the entire application.

Reply:
Really good points, @nkimaina . I'm in agreement with letting organizations create and maintain release branches according to their needs.

I'm also going to propose some changes to our definition of major and minor versions, which might help accommodate the needs you are designing for.